[Clang][CodeGen][X86] don't coerce int128 into `{i64,i64}` for SysV-like ABIs

T0b1-iOS opened the pull request:
Currently, clang coerces (u)int128_t to two i64 IR parameters when they are passed in registers. This leads to broken debug info for them after applying SROA+InstCombine. SROA generates IR like this ([godbolt](https://godbolt.org/z/YrTa4chfc)):
```llvm
define dso_local { i64, i64 } @add(i64 noundef %a.coerce0, i64 noundef %a.coerce1)  {
entry:
  %a.sroa.2.0.insert.ext = zext i64 %a.coerce1 to i128
  %a.sroa.2.0.insert.shift = shl nuw i128 %a.sroa.2.0.insert.ext, 64
  %a.sroa.0.0.insert.ext = zext i64 %a.coerce0 to i128
  %a.sroa.0.0.insert.insert = or i128 %a.sroa.2.0.insert.shift, %a.sroa.0.0.insert.ext
    #dbg_value(i128 %a.sroa.0.0.insert.insert, !17, !DIExpression(), !18)
// ...
!17 = !DILocalVariable(name: "a", arg: 1, scope: !10, file: !11, line: 1, type: !14)
// ...
```
  
  and InstCombine then removes the `or`, moving it into the `DIExpression`, and the `shl` at which point the debug info salvaging in `Transforms/Local` replaces the arguments with `poison` as it does not allow constants larger than 64 bit in `DIExpression`s.
  
  I'm working under the assumption that there is interest in fixing this. If not, please tell me.
  By not coercing `int128_t`s into `{i64, i64}` but keeping them as `i128`, the debug info stays intact and SelectionDAG then generates two `DW_OP_LLVM_fragment` expressions for the two corresponding argument registers.

Given that the ABI code for x64 seems to not coerce the argument when it is passed on the stack, it should not lead to any problems keeping it as an `i128` when it is passed in registers.

Alternatively, this could be fixed by checking if a constant value fits in 64 bits in the debug info salvaging code and then extending the value on the expression stack to the necessary width. This fixes InstCombine breaking the debug info but then SelectionDAG removes the expression and that seems significantly more complex to debug.

Another fix may be to generate `DW_OP_LLVM_fragment` expressions when removing the `or` as it gets marked as disjoint by InstCombine. However, I don't know if the KnownBits information is still available at the time the `or` gets removed and it would probably require refactoring of the debug info salvaging code as that currently only seems to replace single expressions and is not designed to support generating new debug records.

Converting `(u)int128_t` arguments to `i128` in the IR seems like the simpler solution, if it doesn't cause any ABI issues.

tagging potential reviewers: @nikic @phoebewang @rnk

nikic reviewed (commented):
I haven't checked the implementation, but conceptually this looks fine. Passing as i128 instead of {i64, i64} is better for optimization purposes. Rust already passes these as i128, so we know that works, and the change will improve x-lang LTO. (Clang also passes as i128 already, but only for ones that are known to be passed in memory -- go figure.)

efriedma-quic commented:
There isn't any reason we can't do this... but it doesn't really solve the issue in general.  And there are a lot of other ways to trip over this.  CC @dblaike @hvdijk @jmorse 

Do we want to try to handle _BitInt, or a struct containing an i128?

hvdijk commented:
Both agreed with @nikic that this PR is a good idea and with @efriedma-quic that this does not solve the problem in general. I don't think the latter should stop this from being merged but it would be nice to find something that addresses all cases.

> Another fix may be to generate `DW_OP_LLVM_fragment` expressions when removing the `or`

Maybe a bad suggestion, I have not looked into the implementation, but at the point where the `i128` gets split into two `i64`s, could we already update the `#dbg_value` there to use fragments? At that point we have all the required information, and it sounds like it would avoid the problem since there is then no longer any larger-than-`i64` debug info to cause `poison`?

phoebewang commented on clang/lib/CodeGen/Targets/X86.cpp:
```diff
@@ -2595,6 +2595,14 @@ GetX86_64ByValArgumentPair(llvm::Type *Lo, llvm::Type *Hi,
+
   // AMD64-ABI 3.2.3p4: Rule 1. Classify the return type with the
   // classification algorithm.
   X86_64ABIInfo::Class Lo, Hi;
```
Probably remove the handling for int128 in `classify`. It's better to use `classify` as the single entry, but I can't get any idea.

T0b1-iOS replied:
I suppose you could make GetINTEGERTypeAtOffset return an i128 if it sees an i128 as the IRType and IROffset is 0 but then you need to check if it did that in classifyArgumentType and return immediately or make GetINTEGERTypeAtOffset return nullptr if the IROffset is >0 to prevent it from creating an aggregate argument? The latter seems like something one could easily trip over though

efriedma-quic replied:
Something like that, yes.  Or maybe it's simpler to do it in a separate function; whatever's more readable.

T0b1-iOS replied:
I changed GetINTEGERTypeAtOffset to return an i128 when appropriate. That also causes _BitInt(128) to be passed as an i128 but that should be fine as long as it doesn't change how it is passed on the stack.

T0b1-iOS replied:
@efriedma-quic ping. Would this be okay now?

T0b1-iOS replied:
@efriedma-quic ping

efriedma-quic replied:
Sorry about the delay here.

I think this does the right thing, but in a pretty unintuitive way; I had to read through the code multiple times to convince myself this actually works the way it's supposed to.

I'd rather not have GetINTEGERTypeAtOffset return nullptr; instead, the callers of GetINTEGERTypeAtOffset should explicitly check for an i128 return type, and return early.  And assert `Hi == Integer`.

T0b1-iOS replied:
@efriedma-quic Okay, the CI problems are fixed now. It seems there was some external problem with libcxx that made its tests fail before. Maybe you could take a look now

jmorse commented:
There's precedent for splitting the debug-info into fragments early, we do something similar for C++ structured-bindings in SROA and a very similar process happens when the i128 argument is split up in instruction-selection.

Split-up locations become brittle to optimisations though: any source-expression that gets salvaged into a DIExpression can't be decomposed into two fragments, and we have to mark them optimised out. Any salvaging into DIExpression for an i128 will almost certainly fail when the value gets split up in isel. Plus I believe DWARF doesn't support values on the expression stack larger than the machine-address-size, but that's an assumption that can be bent.

Presumably the desire here is to improve the variable availability of parameters rather than fully supporting i128's through LLVM debug-info? If so, the least invasive approach may be to teach SROA to decompose into fragments similar to structured-bindings for the first variable location in the function. Or, passing an i128 instead works alright, I have no opinion on ABI matters.

(Paging @OCHyams as he worked on structured bindings; and @slinder1 as this is another scenario that might be easier with a typed expression stack)

T0b1-iOS commented:
> Maybe a bad suggestion, I have not looked into the implementation, but at the point where the i128 gets split into two i64s, could we already update the #dbg_value there to use fragments? At that point we have all the required information, and it sounds like it would avoid the problem since there is then no longer any larger-than-i64 debug info to cause poison?

I think would be possible but currently the code that inserts the `#dbg_declare` records (`CodeGenFunction::EmitParmDecl`) is oblivious to the splitting. It only gets handed the `i128` value that is loaded from the temporary `{i64, i64}` alloca (created in `CodeGenFunction::EmitFunctionProlog`) where both `i64` arguments are stored into. Thus you would need to add debug handling code into the part that creates the temporary alloca which would require a refactoring of the way debug info metadata is created because the code responsible for that currently creates the `#dbg_declare` records together with the metadata (c.f. `CGDebugInfo::EmitDeclare`).
Passing as i128 seems like the better choice in that case.

To solve this problem in general, it would probably(?) be best to make SROA create fragment `#dbg_value` records if it decomposes an alloca into multiple IR values but I have no idea how difficult that would be to accomplish.

> Presumably the desire here is to improve the variable availability of parameters rather than fully supporting i128's through LLVM debug-info

Yes, at least my primary motivation is for me to be able to retrieve the registers the parameter is passed in without having to rely on any crude heuristics.

jmorse commented:
> To solve this problem in general, it would probably(?) be best to make SROA create fragment #dbg_value records if it decomposes an alloca into multiple IR values but I have no idea how difficult that would be to accomplish.

An example of this happening today with C++ structured bindings is here: https://godbolt.org/z/rn3os13M7 . Specifically, the i64 return value from "f", a squashed-together pair of i32s, is decomposed into two `#dbg_value`s corresponding to the two variables that the structured-binding assigns to. Each `#dbg_value` refers to different IR Values, which seems similar to what's desired here. The major difference is that structured bindings have independent source variables contained within one alloca; wheras the situation here is one source variable being split into fragments.

T0b1-iOS committed:
[Clang][CodeGen][X86] don't coerce int128 into `{i64,i64}` for SysV-like ABIs

T0b1-iOS committed:
Always return i128 from GetINTEGERTypeAtOffset and early return in callers

T0b1-iOS committed:
fix clang-format

T0b1-iOS committed:
[Clang][CodeGen][X86] handle (u)int128_t in structs.

As a side-effect, this will also pass _BitInt(128) as a i128 if it is
passed in registers

efriedma-quic reviewed (approved):
LGTM with one minor suggestion

T0b1-iOS committed:
Update clang/lib/CodeGen/Targets/X86.cpp

Co-authored-by: Eli Friedman <efriedma@quicinc.com>

T0b1-iOS commented:
Fixed. Thanks for the review! As I don't have commit access, could you merge this for me?. And do you think this could be put onto the 21 release branch? I think it should be possible from a policy pov but I think you need to have commit access to make a backport request if I'm reading the docs correctly.
